src/main/java/org/embulk/output/mailchimp/MailChimpRecordBuffer.java

```diff
@@ -206,26 +208,38 @@ public JsonNode apply(JsonNode input)
 
                 // Update additional merge fields if exist
                 if (task.getMergeFields().isPresent() && !task.getMergeFields().get().isEmpty()) {
-                    for (final Column column : schema.getColumns()) {
-                        if (!"".equals(containsCaseInsensitive(column.getName(), task.getMergeFields().get()))) {
-                            String value = input.hasNonNull(column.getName()) ? input.findValue(column.getName()).asText() : "";
-
-                            // Try to convert to Json from string with the merge field's type is address
-                            if (availableMergeFields.get(column.getName().toLowerCase()).getType()
-                                    .equals(MergeField.MergeFieldType.ADDRESS.getType())) {
-                                JsonNode addressNode = toJsonNode(value);
-                                if (addressNode instanceof NullNode) {
-                                    mergeFields.put(column.getName().toUpperCase(), value);
-                                }
-                                else {
-                                    mergeFields.set(column.getName().toUpperCase(),
-                                                    orderJsonNode(addressNode, AddressMergeFieldAttribute.values()));
-                                }
+                    Map<String, String> columnNameLookup = new TreeMap<>(CASE_INSENSITIVE_ORDER);
```

Comment:
Should we consider using the HashSet ? Because I see it is not necessary to use the Map

Reply:
The intent is to have a mapping from case-insensitive column name to the original column name. So the Task's merge field can be compare/match with the supplied columns, but still retaining the original column name so the value can be retrieved from input with that original column name. That's why it need to be a _map_.
String's hashing yield different values on different character's cases (apparently different char code), that's why it need to be a _tree_ implementation with an explicit case-insensitive comparator instead of hashes.

Reply:
Ok I got it. Thank you for clarifying